rclcpp_lifecycle/src/change_state_handler_impl.cpp

```diff
@@ -0,0 +1,335 @@
+    // Update current state
+    current_state_ = State(state_machine_.current_state);
+  }
+  is_transitioning_.store(false);
+}
+
+bool ChangeStateHandlerImpl::is_transitioning() const
+{
+  return is_transitioning_.load();
+}
+
+bool
+ChangeStateHandlerImpl::is_async_callback(
+  unsigned int cb_id) const
+{
+  return async_cb_map_.find(static_cast<uint8_t>(cb_id)) != async_cb_map_.end();
+}
+
+node_interfaces::LifecycleNodeInterface::CallbackReturn
+ChangeStateHandlerImpl::execute_callback(
+  unsigned int cb_id, const State & previous_state) const
+{
+  // in case no callback was attached, we forward directly
+  auto cb_success = node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+
+  auto it = cb_map_.find(static_cast<uint8_t>(cb_id));
+  if (it != cb_map_.end()) {
+    auto callback = it->second;
+    try {
+      cb_success = callback(State(previous_state));
+    } catch (const std::exception & e) {
+      RCUTILS_LOG_ERROR("Caught exception in callback for transition %d", it->first);
+      RCUTILS_LOG_ERROR("Original error: %s", e.what());
+      cb_success = node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR;
+    }
+  }
+  return cb_success;
+}
+
+void
+ChangeStateHandlerImpl::execute_async_callback(
+  unsigned int cb_id,
+  const State & previous_state)
+{
+  auto it = async_cb_map_.find(static_cast<uint8_t>(cb_id));
+  if (it != async_cb_map_.end()) {
+    auto callback = it->second;
+    try {
+      callback(State(previous_state), shared_from_this());
+    } catch (const std::exception & e) {
+      RCUTILS_LOG_ERROR("Caught exception in callback for transition %d", it->first);
+      RCUTILS_LOG_ERROR("Original error: %s", e.what());
+      send_callback_resp(
+        node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR);
+    }
+  } else {
+    // Note that is_async_callback should be called before execute_async_callback
+    // therefor this should never run
+    // However, to be consistent with execute_callback:
+    // in case no callback was attached, we forward directly
+    send_callback_resp(
+      node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS);
+  }
+}
+
+const char *
+ChangeStateHandlerImpl::get_label_for_return_code(
+  node_interfaces::LifecycleNodeInterface::CallbackReturn cb_return_code)
+{
+  auto cb_id = static_cast<uint8_t>(cb_return_code);
+  if (cb_id == lifecycle_msgs::msg::Transition::TRANSITION_CALLBACK_SUCCESS) {
+    return rcl_lifecycle_transition_success_label;
+  } else if (cb_id == lifecycle_msgs::msg::Transition::TRANSITION_CALLBACK_FAILURE) {
+    return rcl_lifecycle_transition_failure_label;
+  }
+  return rcl_lifecycle_transition_error_label;
+}
+
+void
+ChangeStateHandlerImpl::rcl_ret_error()
+{
+  finalize_change_state(false);
+}
+
+bool
+ChangeStateHandlerImpl::in_non_error_transition_state() const
+{
+  std::lock_guard<std::recursive_mutex> lock(state_machine_mutex_);
```

Comment:
This function and `in_error_transition_state()` both lock the `state_machine_mutex_` in order to get the `current_state_id`.

I think this should be pulled out into its own method `get_current_state_id()` so we are not locking twice within [send_callback_resp](https://github.com/tgroechel/rclcpp/pull/1/files#diff-f10f2a1749e1b2eabcfe19d560d830d0a99b8754121a51e9f8479ca459bc823fR64) and then have these take in a `uint8_t state_id`